lighthouse-cli/chrome-launcher.js

```diff
@@ -0,0 +1,175 @@
+const mkdirp = require('mkdirp');
+const net = require('net');
+const rimraf = require('rimraf');
+const ask = require('./ask');
+const chromeFinder = require('./chrome-finder');
+
+const spawn = childProcess.spawn;
+const execSync = childProcess.execSync;
+
+module.exports = class Launcher {
+  constructor(opts) {
+    opts = opts || {};
+    // choose the first one (default)
+    this.head = defaults(opts.head, true);
+    this.pollInterval = 500;
+    this.chromeInstances = [];
+    this.prepared = false;
+  }
+
+  flags() {
+    return [
+      '--remote-debugging-port=9222',
+      '--no-first-run',
+      `--user-data-dir=${this.TMP_PROFILE_DIR}`
+    ];
+  }
+
+  prepare() {
+    switch (process.platform) {
+      case 'darwin':
+        this.TMP_PROFILE_DIR = unixTmpDir();
+        break;
+
+      case 'linux':
+        this.TMP_PROFILE_DIR = unixTmpDir();
+        break;
+
+      case 'win32':
+        this.TMP_PROFILE_DIR = win32TmpDir();
+        break;
+
+      default:
+        throw new Error('Platform ' + process.platform + ' is not supported');
+    }
+
+    this.outFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome-out.log`, 'a');
+    this.errFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome-err.log`, 'a');
+    this.pidFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome.pid`, 'w');
+
+    console.log(`created ${this.TMP_PROFILE_DIR}`);
+
+    this.prepared = true;
+  }
+
+  run() {
+    if (!this.prepared) {
+      this.prepare();
+    }
+
+    return Promise.resolve()
+      .then(() => {
+        const installations = chromeFinder[process.platform]();
+
+        switch (true) {
+          case installations.length < 1:
+            return Promise.reject(new Error('No Chrome Installations Found'));
+
+          case installations.length === 1:
+            return installations[0];
+
+          case this.head:
+            return installations[0];
+
+          default:
+            return ask('Choose a Chrome installation to use with Lighthouse', installations);
+        }
+      })
+      .then(execPath => this.spawn(execPath));
+  }
+
+  spawn(execPath) {
+    return new Promise((resolve, reject) => {
+      const chrome = spawn(
+        execPath,
+        this.flags(),
+        {
+          detached: true,
+          stdio: ['ignore', this.outFile, this.errFile]
+        }
+      );
+      this.chromeInstances.push(chrome);
+
+      fs.writeFileSync(this.pidFile, chrome.pid.toString());
+
+      console.log('Chrome running with pid =', chrome.pid);
+      resolve(chrome.pid);
+    })
+    .then(pid => Promise.all([pid, this.poll(0)]));
+  }
+
+  cleanup(client) {
+    if (client) {
+      client.removeAllListeners();
+      client.end();
+      client.destroy();
+      client.unref();
+    }
+  }
+
+  connect() {
+    return new Promise((resolve, reject) => {
+      const client = net.createConnection(9222);
+      client.once('error', err => {
+        this.cleanup(client);
+        reject(err);
```

Comment:
this connect function is used only after chrome is spawned ... Hmmm... maybe I can just check before anything starts - and revert the logic - try to launch chrome first and then do the lighthouse run .... in trying to launch chrome if 9222 is open, skip launching and go directly to lighthouse run.